crates/circuit/src/dag_circuit.rs

```diff
@@ -2771,43 +2773,121 @@ def _format(operand):
+                        }
+                        NodeType::QubitIn(q) => {
+                            let qbit_in = new_dag.qubit_input_map.get(q).unwrap();
+                            node_map.insert(*node, *qbit_in);
+                            non_classical = true;
+                        }
+                        NodeType::QubitOut(q) => {
+                            let qbit_out = new_dag.qubit_output_map.get(q).unwrap();
+                            node_map.insert(*node, *qbit_out);
+                            non_classical = true;
+                        }
+                        NodeType::Operation(pi) => {
+                            let qubits = self.qargs_cache.intern(pi.qubits_id);
+                            let clbits = self.cargs_cache.intern(pi.clbits_id);
+                            let qubits_id =
+                                Interner::intern(&mut new_dag.qargs_cache, qubits.clone())?;
+                            let clbits_id =
+                                Interner::intern(&mut new_dag.cargs_cache, clbits.clone())?;
+
+                            let extr_attr = pi.extra_attrs.clone().unwrap();
+                            let new_pi = NodeType::Operation(PackedInstruction::new(
+                                pi.op.clone(),
+                                qubits_id,
+                                clbits_id,
+                                pi.params.clone(),
+                                extr_attr.label,
+                                extr_attr.duration,
+                                extr_attr.unit,
+                                extr_attr.condition,
+                                #[cfg(feature = "cache_pygates")]
+                                pi.py_op.borrow().clone(),
+                            ));
+                            let new_node = new_dag.dag.add_node(new_pi);
+                            node_map.insert(*node, new_node);
+                            non_classical = true;
+                        }
+                    },
+                    None => panic!("DAG node without payload!"),
+                }
+            }
+            if !non_classical {
+                continue;
+            }
+
+            // Handling the edges in the new dag
+            for node in comp_nodes {
+                // Since the nodes comprise an SCC, it's enough to just look at the (e.g.) outgoing edges
+                let outgoing_edges = self.dag.edges_directed(*node, Direction::Outgoing);
+
+                // Remove the edges added by copy_empty_like (as idle wires) to avoid duplication
+                if let Some(NodeType::QubitIn(_)) | Some(NodeType::ClbitIn(_)) =
+                    self.dag.node_weight(*node)
+                {
+                    let edges: Vec<EdgeIndex> = new_dag.dag.edges(*node).map(|e| e.id()).collect();
+                    for edge in edges {
+                        new_dag.dag.remove_edge(edge);
+                    }
+                }
```

Comment:
It might be better to do this as one step before this loop (remove all of the graph's edges at once). It would be a little clearer that we don't want to start off with any edges, and it would also save `Vec` allocations (we'd just do one `collect` up front, rather than a `collect` for every bit).